src/main/scala/example/repository/UserRepositoryFuture.scala

```diff
@@ -2,18 +2,34 @@ package example.repository
 
 import java.util.concurrent.atomic.AtomicLong
 import example.domain.User
+import example.schema.UserTable
 import scala.concurrent.{ExecutionContext, Future}
+import slick.jdbc.H2Profile.api._
+import slick.lifted.TableQuery
 
-class UserRepositoryFuture(implicit val ec: ExecutionContext) extends UserRepository[Future] {
-  private var storage: Map[Long, User] = Map()
+/**
+  * Future implementation of the User repository.
+  * @param db the injected database to manage operations into.
+  * @param ec the scope of execution context used by the repository.
+  */
+class UserRepositoryFuture(db: Database)(implicit val ec: ExecutionContext) extends UserRepository[Future] {
+
+  lazy val users = TableQuery[UserTable]
   private val id = new AtomicLong(0)
 
-  override def registerUser(username: String): Future[User] = Future {
+  override def registerUser(username: String, address: Option[String], email: String): Future[User] = Future {
     val userId = id.incrementAndGet()
-    val user = User(userId, username)
-    storage = storage + (userId -> user)
+    val user = User(userId, username, address, email)
+    db.run(users.insertOrUpdate(user))
     user
   }
-  override def getById(id: Long): Future[Option[User]] = Future {storage.get(id)}
-  override def getByUsername(username: String): Future[Option[User]] = Future {storage.values.find(user => user.username == username)}
+  override def getById(id: Long): Future[Option[User]] = {
+    db.run(users.filter(_.id === id).result.headOption)
+  }
```

Comment:
no retries here, you should use slightly modified version of retryAsync which retries some DB SQL exceptions like timeouts, etc.